### DISCUSSION TITLE
Travel speed

### DISCUSSION THREAD

**Discussion by Kosnov** [kachurovskiy/nanoels#207] [Ideas]
Максим, доброго. Скорость движения при нажатии стрелок (влево, вправо, вперед, назад) большая. Можно ли изменить код программы, чтобы задать скорость (кнопками+ и -) перемещения перед движением.

**Comment by Viktor1972-avi** [comment]
Оставить 3 скорости фиксированные и добавит одну произвольную на усмотрение оператора, с возможность оперативного изменения + -.

**Comment by Kosnov** [comment]
Может кто-нибудь сделает(изменит код программы), чтоб при перемещении кнопками(влево, вправо, вперед, назад) движение происходило с той подачей, которая выставлена.

**Comment by duebelpapst** [comment]
Use numpad buttons 0 to 9 to enter a custom step value. Pressing Steps when the screen bottom line shows Use 1.234mm? will make the step equal to 1.234mm.

**Reply by Kosnov** [reply]
no, that's not it

**Reply by Kosnov** [reply]
I use the numeric keypad buttons 0-9, set the Pitch to 0.07, press any arrow - should move at a speed of 0.07

Я использую цифровые кнопки клавиатуры 0-9, устанавливаю Pitch на 0,07, нажимаю любую стрелку - должно двигаться со скоростью 0,07

**Comment by duebelpapst** [comment]
Pressing Steps means pressing the buttom ![image](https://github.com/user-attachments/assets/3f9a520f-aaff-41ee-8c34-a1179d5ba593)

**Comment by Kosnov** [comment]
no, that's not it

**Comment by Kosnov** [comment]
Я не программист, я бы так и сделал. Если уважаемый Автор (Максим К - успехов и здоровья) читает эту ветку, то он понимает, о чем я говорю.

**Reply by kachurovskiy** [reply]
И вам всех благ. Если вы хотите, чтобы при ручном перемещении использовалась скорость pitch/sec, можно попробовать этого добиться модифицировав вот эту строку - https://github.com/kachurovskiy/nanoels/blob/main/h4/h4.ino#L1126

Сейчас там написано `return isContinuousStep() ? a->speedManualMove : ...`

Соответственно нам нужно вместо `a->speedManualMove` поставить шаги в секунду как указано в `dupr`. Проблема в том, что дупр имеет размерность деци-микрон а нам нужна размерность шаг / секунда. Поэтому укажем `abs(dupr) * a->motorSteps / a->screwPitch` который даёт размерность шаг а секунду возьмем за 1 а дупр возьмем по модулю т.к. он может быть негативным. Еще он может быть нулём.

Итого можно попробовать заменить строку 1126 на следущее и может что-то получится - правда совершенно не уверен что это будет удобно:

```
  if (!isContinuousStep()) return min(long(a->speedManualMove), abs(getMoveStepForAxis(a)) * 1000 / STEP_TIME_MS);
  if (dupr == 0) return a->speedManualMove;
  return abs(dupr) * a->motorSteps / a->screwPitch;
```

**Comment by Kosnov** [comment]
Благодарю, проверю-отпишусь.

**Comment by Kosnov** [comment]
Максим доброго. Прошил, да, скорость перемещения меняется, но нажав и удерживая например кнопку влево- перемещение происходит на длину 1мм, дальше остановка.
Надо чтоб например- удерживая кнопку влево происходило перемещение, отпустив кнопку- остановка

**Reply by kachurovskiy** [reply]
Паузы при перемещении происходят если у вас установлен нестандартный шаг. Чтобы не было пауз поставьте ![image](https://github.com/user-attachments/assets/3f9a520f-aaff-41ee-8c34-a1179d5ba593) 1мм.

Вот более правильный вариант функции скорости который вам нужен:

```
long getStepMaxSpeed(Axis* a) {
  if (!isContinuousStep()) return min(long(a->speedManualMove), abs(getMoveStepForAxis(a)) * 1000 / STEP_TIME_MS);
  if (dupr == 0) return a->speedManualMove;
  return max(a->speedStart, (long) round(abs(dupr) * a->motorSteps / a->screwPitch));
}
```

Чтобы достигать достаточно маленьких скоростей ручного перемещения уменьшите `SPEED_START_Z` и `SPEED_START_X` например в 2-4 раза.

**Comment by Kosnov** [comment]
Доброго Максим. Ничего не выходит. Режим шестеренок. Step 1mm. Устанавливаю Pitch 0,01-0,5 перемещение стрелками с одинаковой скоростью.

**Reply by kachurovskiy** [reply]
На таком маленьком питче перемещение происходит с минимальной скоростью SPEED_START_Z / SPEED_START_X. Уменьшите эти значения ещё чтобы ощутить изменения скорости перемещения на минимальных питчах или повысьте питч до 2-3мм.

**Comment by Kosnov** [comment]
const long MAX_TRAVEL_MM_Z = 300; // Lathe bed doesn't allow to travel more than this in one go, 30cm / ~1 foot
Что означает эта строчка, для чего?

**Reply by kachurovskiy** [reply]
Чтобы аварийно останавливаться если контроллер запрашивает перемещение на длину более данной что физически невозможно и следовательно баг в программе или ошибка оператора.

**Comment by Kosnov** [comment]
Основные подачи у меня от 0,07 до 0,15
SPEED_START_Z -что тут уменьшить - const long SPEED_START_Z = 2 * MOTOR_STEPS_Z;

**Reply by kachurovskiy** [reply]
Уменьшайте `SPEED_START_Z` до тех пор пока скорость ручного перемещения не будет устраивать. Попробуйте значения 100, 50, 20.

**Comment by Kosnov** [comment]
const long SPEED_START_Z = 2 * MOTOR_STEPS_Z где здесь поменять на 100,50, 20

**Reply by kachurovskiy** [reply]
Вместо

```
const long SPEED_START_Z = 2 * MOTOR_STEPS_Z; // Initial speed of a motor, steps / second.
```

написать

```
const long SPEED_START_Z = 100; // Initial speed of a motor, steps / second.
```

**Comment by Kosnov** [comment]
Я наверное не могу просто правильно описать то, что хотелось бы видеть в прошивке.
У меня джойстик вместо стрелок.
Установив Pitch 0,07- джойстик влево,вправо, вперед, назад и поехал, джойстик отпустил- остановка, но этого не происходит(скорость одна и та, и доезжает кратно 2мм). У кого есть данная гитара, перешли с Олеговской,со мной согласны.
Может кто нибудь исправит со временем.
Максим, обязуюсь больше не задавать вопросов и не потревожу.

**Reply by kachurovskiy** [reply]
Если мин. скорость 100 не спасает, попробуйте 20.

```
const long SPEED_START_Z = 20;
```

Кратность 1мм идёт от того, что шаг установлен 1мм. Скачет на 2мм обычно т.к. быстро отпустить кнопку сложно. Если хотите более мелкий стандартный шаг, это можно поменять в строке 196, например:

```
#define MOVE_STEP_1 2000 // 0.2mm
```
